File: examples/F#/OpenGL/Tutorial 1.3 - Shaders/shader.fs
module Shader

open System
open Silk.NET.OpenGL

// Used to load and compile the vertex and fragment shaders
let private createShader (gl: GL) (shaderType: GLEnum) path =
    let shader = gl.CreateShader(shaderType)
    let source = IO.File.ReadAllText(path)

    // Load and compile the shader
    gl.ShaderSource(shader, source)
    gl.CompileShader(shader)

    // Throw an exception if compiling the shader failed
    // The info log will tell us what line the shader failed on
    // so we can inspect the source.
    if gl.GetShader(shader, GLEnum.CompileStatus) <> 1 then
        let info = gl.GetShaderInfoLog(shader)
        gl.DeleteShader(shader)
        failwith (sprintf "Error creating shader: %A" info)

    shader


let create (gl: GL) =
    // Load and compile our vertex and fragment shaders
    let vertexShader = createShader gl GLEnum.VertexShader "shader.vert"
    let fragmentShader = createShader gl GLEnum.FragmentShader "shader.frag"

    // Link the vertex and fragment shaders into a shader program
    let shaderProgram = gl.CreateProgram()
    gl.AttachShader(shaderProgram, vertexShader)
    gl.AttachShader(shaderProgram, fragmentShader)
    gl.LinkProgram(shaderProgram)
    
    // We're done with the shaders so they can be removed to free memory
    gl.DetachShader(shaderProgram, fragmentShader)
    gl.DetachShader(shaderProgram, vertexShader)
    gl.DeleteShader(vertexShader)
    gl.DeleteShader(fragmentShader)

    // Throw an exception if linking has failed
    if gl.GetProgram(shaderProgram, GLEnum.LinkStatus) <> 1 then
        let info = gl.GetProgramInfoLog(shaderProgram)
        failwith (sprintf "Error linking program %A" info)

    shaderProgram

File: examples/CSharp/Tutorial 1.3 - Abstractions/Shader.cs
using System;
using System.IO;
using Silk.NET.OpenGL;

namespace Tutorial
{
    public class Shader : IDisposable
    {
        //Our handle and the GL instance this class will use, these are private because they have no reason to be public.
        //Most of the time you would want to abstract items to make things like this invisible.
        private uint _handle;
        private GL _gl;

        public Shader(GL gl, string vertexPath, string fragmentPath)
        {
            _gl = gl;

            //Load the individual shaders.
            uint vertex = LoadShader(ShaderType.VertexShader, vertexPath);
            uint fragment = LoadShader(ShaderType.FragmentShader, fragmentPath);
            //Create the shader program.
            _handle = _gl.CreateProgram();
            //Attach the individual shaders.
            _gl.AttachShader(_handle, vertex);
            _gl.AttachShader(_handle, fragment);
            _gl.LinkProgram(_handle);
            //Check for linking errors.
            _gl.GetProgram(_handle, GLEnum.LinkStatus, out var status);
            if (status == 0)
            {
                throw new Exception($"Program failed to link with error: {_gl.GetProgramInfoLog(_handle)}");
            }
            //Detach and delete the shaders
            _gl.DetachShader(_handle, vertex);
            _gl.DetachShader(_handle, fragment);
            _gl.DeleteShader(vertex);
            _gl.DeleteShader(fragment);
        }

        public void Use()
        {
            //Using the program
            _gl.UseProgram(_handle);
        }

        //Uniforms are properties that applies to the entire geometry
        public void SetUniform(string name, int value)
        {
            //Setting a uniform on a shader using a name.
            int location = _gl.GetUniformLocation(_handle, name);
            if (location == -1) //If GetUniformLocation returns -1 the uniform is not found.
            {
                throw new Exception($"{name} uniform not found on shader.");
            }
            _gl.Uniform1(location, value);
        }

        public void SetUniform(string name, float value)
        {
            int location = _gl.GetUniformLocation(_handle, name);
            if (location == -1)
            {
                throw new Exception($"{name} uniform not found on shader.");
            }
            _gl.Uniform1(location, value);
        }

        public void Dispose()
        {
            //Remember to delete the program when we are done.
            _gl.DeleteProgram(_handle);
        }

        private uint LoadShader(ShaderType type, string path)
        {
            //To load a single shader we need to:
            //1) Load the shader from a file.
            //2) Create the handle.
            //3) Upload the source to opengl.
            //4) Compile the shader.
            //5) Check for errors.
            string src = File.ReadAllText(path);
            uint handle = _gl.CreateShader(type);
            _gl.ShaderSource(handle, src);
            _gl.CompileShader(handle);
            string infoLog = _gl.GetShaderInfoLog(handle);
            if (!string.IsNullOrWhiteSpace(infoLog))
            {
                throw new Exception($"Error compiling shader of type {type}, failed with error {infoLog}");
            }

            return handle;
        }
    }
}
